### PR TITLE
DefineNamedBindings: use a proxy Statement rather than inspecting for `NullArgument`

### LINKED ISSUE #1416: Improve/fix DefineBinds nullness detection
Current `instanceof NullArgument` is rather fickle. ATM it doesn't work for some arguments, like `OptionalArgumentFactory` and `InputStreamArgument`. @stevenschlansker appears to be taking care of it for `ObjectArgument` in #1409.

Suggestion:
```
@FunctionalInterface
public interface Argument {
    /**
     * @return true if the value of this binding is null, false otherwise
     */
    default boolean isNull() {
        return false;
    }
}
```
Override in `NullArgument` with `return true` and in others with a `== null` on a value or whatever's appropriate. `OptionalArgumentFactory` for example would do `isEmpty()`. This is also more discoverable by argument implementors than a javadoc and manual comment, I think.

### PR DESCRIPTION
**Pull request by stevenschlansker**
Fixes #1416

### CODE REVIEW

**Comment by stevenschlansker** [comment]
rebased and squished

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java [hunk 1]
```diff
@@ -13,20 +13,50 @@
...
+                boolean argNull = args.length > 1 && args[1] == null;
+                setNull = method.getName().equals("setNull") || argNull;
+            }
+            return null;
+        }
+
+        void define(String name, Argument a, StatementContext ctx) {
+            setNull = false;
+            setCalled = false;
+            Unchecked.runnable(() -> a.apply(1, fakeStmt, ctx)).run();
```
[the remark this answers is not in the document]
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: I guess.  I'm not sure that I agree that having two different approaches to passing through the three arguments is clearer, though.  It reads fine either way to me.  But I guess we don't need runnable support at all if we change it.

[the remark this answers is not in the document]
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: And it trips up the compiler too.  Now it's

```java
Unchecked.<PreparedStatement>consumer(stmt -> a.apply(1, stmt, ctx)).accept(fakeStmt);
```

Are you sure that's better?  It's significantly longer and introduces two additional constructs that weren't there in the first place (the `<>` and the threading-through of `fakeStmt -> stmt -> apply`).

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java [hunk 2]
```diff
@@ -13,20 +13,50 @@
...
         final Binding binding = ctx.getBinding();
+        final Handler handler = new Handler();
         binding.getNames().stream()
             .filter(name -> !alreadyDefined.contains(name))
             .forEach(name -> binding.findForName(name, ctx).ifPresent(
-                    a -> ctx.define(name, a instanceof NullArgument ? false : true)));
+                    a -> handler.define(name, a, ctx)));
+    }
+
+    static class Handler implements InvocationHandler {
```
[the remark this answers is not in the document]
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: Isn't its purpose clear from context?  It's not intended to be used from an external location.  And the term 'handler' is what it is actually doing.

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java [hunk 3]
```diff
@@ -13,20 +13,50 @@
...
+        private final PreparedStatement fakeStmt = (PreparedStatement)
+                Proxy.newProxyInstance(Thread.currentThread().getContextClassLoader(), new Class<?>[] {PreparedStatement.class}, this);
+        private boolean setNull;
+        private boolean setCalled;
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            if (setCalled = method.getName().startsWith("set")) { // NOPMD
+                boolean argNull = args.length > 1 && args[1] == null;
+                setNull = method.getName().equals("setNull") || argNull;
```
[the remark this answers is not in the document]
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: I guess.  but the method name is never null...

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: A rose by any other name is still a rose, but in java we name classes after their purpose anyway :) `SetNullDetectingHandler` sound good to you then?

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: I guess thats a java 11 thing cause it didn't happen to me. I agree with you then

[the remark this answers is not in the document]
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: hm, ok, I'll push it and lets see what Travis says

[the remark this answers is not in the document]
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: The spurious failures are super annoying.

[the remark this answers is not in the document]
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: java8 travis fails too:
> [ERROR] /home/travis/build/jdbi/jdbi/core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java:[54,53] incompatible types: java.lang.Object cannot be converted to java.sql.PreparedStatement

File: core/src/main/java/org/jdbi/v3/core/internal/exceptions/Unchecked.java [hunk 4]
```diff
@@ -38,6 +38,17 @@ private Unchecked() {
         };
     }
 
+    @SuppressWarnings("PMD.DoNotUseThreads")
```
Comment (TheRealMarnes) [inline core/src/main/java/org/jdbi/v3/core/internal/exceptions/Unchecked.java]: lol

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java [hunk 5]
```diff
@@ -13,20 +13,48 @@
...
+                Proxy.newProxyInstance(Thread.currentThread().getContextClassLoader(), new Class<?>[] {PreparedStatement.class}, this);
+        private boolean setNull;
+        private boolean setCalled;
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) {
+            setCalled = method.getName().startsWith("set");
+            boolean argNull = args.length > 1 && args[1] == null;
+            setNull = "setNull".equals(method.getName()) || argNull;
+            return null;
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: The only thing I can think that _might_ be a problem is that any argument implementations that depend on e.g. `statement.isWrapperFor()` or `statement.unwrap()` to unwrap a vendor-specific prepared statement might error with `@DefineNamedBindings` where otherwise it might behave fine.

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java [hunk 6]
```diff
@@ -13,20 +13,48 @@
...
+
+    private static class SetNullHandler implements InvocationHandler {
+        private final PreparedStatement fakeStmt = (PreparedStatement)
+                Proxy.newProxyInstance(Thread.currentThread().getContextClassLoader(), new Class<?>[] {PreparedStatement.class}, this);
+        private boolean setNull;
+        private boolean setCalled;
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) {
+            setCalled = method.getName().startsWith("set");
```
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: Might not be a single method call. Should probably change this to `||=`

Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: Something like this, rather:

```java
if (method.getName().startsWith("set")) {
  setCalled = true;
  ...etc
}
```

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java (hunk 5 above)
Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: Should return a zero / false value for primitive return types, otherwise ClassCastException

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java (hunk 6 above)
Comment (stevenschlansker) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: I'd personally go for `|=` to save the `if` block but the logic change makes sense.

Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: I think it needs to be in an if-block, otherwise the test for null on the following lines isn't valid.

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java (hunk 5 above)
Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: The `statement.unwrap()` idea isn't going to work, since all the types you would want to unwrap are classes, not interfaces.

Best we can do is provide an informative exception message stating that because `@DefineNamedBindings` uses proxies to determine which fields are non-null, it is incompatible with arguments that rely on `Wrapper.unwrap()`

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java (hunk 6 above)
Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: Pushed fix

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java (hunk 5 above)
Comment (qualidafial) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: Pushed fix

File: core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java [hunk 7]
```diff
@@ -13,20 +13,89 @@
...
+                setCalled = true;
+                boolean argNull = args.length > 1 && args[1] == null;
+                setNull = argNull || "setNull".equals(method.getName());
+            }
+
+            return defaultValue(method.getReturnType());
+        }
+
+        private Object defaultValue(Class<?> type) {
+            if (type.isPrimitive()) {
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: It'd be possible to shorten this significantly: `Array.get(Array.newInstance(type, 1), 0)` but maybe that's too obscure

Comment (TheRealMarnes) [reply core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java]: This smells of a Map<Class, Object> to me
